plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/server/ModuleUtils.java

```diff
@@ -56,4 +60,27 @@ public static String getServiceId(IModule module) {
     
     return "default";
   }
+
+  /**
+   * Returns the set of all referenced modules, including child modules. This returns the unique
+   * modules, and doesn't return the module paths.
+   */
+  public static IModule[] getAllModules(IServer server) {
+    Set<IModule> modules = new LinkedHashSet<>();
+    for (IModule module : server.getModules()) {
+      modules.add(module);
+      addChildModules(server, new IModule[] {module}, modules);
+    }
+    return modules.toArray(new IModule[modules.size()]);
+  }
+
+  /** Recursively walk the children from {@code modulePath}. */
+  private static void addChildModules(IServer server, IModule[] modulePath, Set<IModule> modules) {
+    IModule[] newModulePath = Arrays.copyOf(modulePath, modulePath.length + 1);
+    for (IModule child : server.getChildModules(modulePath, null)) {
```

Comment:
This code follows the same approach used by the `Server#visit()`/`#visitModules()` that is used in its internal `Server#getAllModules()` (unfortunately not exposed via `IServer`).  So it's no worse that what's currently used internally.

I honestly don't expect server module trees to be deep.  But given that we're just accumulating individual modules, we can shortcircuit the descent.